### PR TITLE
#13268: Implement fold

### PR DESCRIPTION
**Pull request by DuongQLee**
### Ticket
Link to Github Issue: [#13268](https://github.com/tenstorrent/tt-metal/issues/13268)

### Problem description
The current fold implementation in tt-metal is incomplete as it doesn't support output_size, kernel_size, and dilation like pytorch fold.

### What's changed
I implemented a new fold operation in moreh_fold to support params exactly like pytorch.
I will clean up old fold implementation in later commits if possible. The current fold implementation is used in some tests and models (Like resnet) so it is hard to clean up.

### Checklist
- [x] Post commit CI passes: https://github.com/tenstorrent/tt-metal/actions/runs/11604992148
- [x] Blackhole Post commit (if applicable): NA
- [x] Model regression CI testing passes (if applicable): NA
- [x] Device performance regression CI testing passes (if applicable): NA
- [x] New/Existing tests provide coverage for changes: 15/15 tests passed

### CODE REVIEW

File: ttnn/cpp/ttnn/operations/moreh/moreh_fold/device/fold_program_factory_rm.cpp [hunk 1]
```diff
@@ -0,0 +1,218 @@
...
+
+using namespace tt;
+using namespace tt::tt_metal;
+
+MorehFoldOperation::ProgramFactory::cached_program_t MorehFoldOperation::ProgramFactory::create(
+    const operation_attributes_t& operation_attributes,
+    const tensor_args_t& tensor_args,
+    tensor_return_value_t& output_tensor) {
+    auto& input = tensor_args.input;
+    auto& output = output_tensor;
```
Comment (thanhnguyen-moreh) [inline ttnn/cpp/ttnn/operations/moreh/moreh_fold/device/fold_program_factory_rm.cpp:22]: I think you should use the output_tensor directly (Btw, change the name to output if you can)

File: ttnn/cpp/ttnn/operations/moreh/moreh_fold/fold_pybind.cpp [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+        "Moreh Fold Operation",
+        ttnn::pybind_arguments_t{
+            py::arg("input"),
+            py::arg("output") = std::nullopt,
+            py::arg("output_size"),
+            py::arg("kernel_size"),
+            py::arg("dilation") = std::vector<uint32_t>{1, 1},
+            py::arg("padding") = std::vector<uint32_t>{0, 0},
+            py::arg("stride") = std::vector<uint32_t>{1, 1},
+            py::arg("output_memory_config") = std::nullopt,
```
Comment (thanhnguyen-moreh) [inline ttnn/cpp/ttnn/operations/moreh/moreh_fold/fold_pybind.cpp:24]: Shouldn't these have kw_only specified if they're optional?

Comment (DuongQLee) [reply ttnn/cpp/ttnn/operations/moreh/moreh_fold/fold_pybind.cpp:24]: This aligns with pytorch implementation. In pytorch description these params are not kw only.

File: ttnn/cpp/ttnn/operations/moreh/moreh_fold/device/fold_device_operation.cpp [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+    auto output_shape = ttnn::SimpleShape(
+        std::vector<uint32_t>{N, C, operation_attributes.output_size[0], operation_attributes.output_size[1]});
+    return output_shape;
+};
+
+MorehFoldOperation::tensor_return_value_t MorehFoldOperation::create_output_tensors(
+    const operation_attributes_t& operation_attributes, const tensor_args_t& tensor_args) {
+    const auto& output_shape = compute_output_shapes(operation_attributes, tensor_args);
+    auto dtype = tensor_args.input.get_dtype();
+    Layout layout{Layout::ROW_MAJOR};
```
[the remark this answers is not in the document]
Comment (DuongQLee) [reply ttnn/cpp/ttnn/operations/moreh/moreh_fold/device/fold_device_operation.cpp:74]: Resolved in latest commit

File: ttnn/cpp/ttnn/operations/moreh/moreh_fold/device/fold_device_operation.hpp [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+            KernelHandle unary_writer_kernel_id;
+            std::vector<CoreCoord> cores;
+        };
+
+        using cached_program_t = ttnn::device_operation::CachedProgram<shared_variables_t>;
+
+        static cached_program_t create(
+            const operation_attributes_t& operation_attributes,
+            const tensor_args_t& tensor_args,
+            tensor_return_value_t& output_tensor);
```
[the remark this answers is not in the document]
Comment (DuongQLee) [reply ttnn/cpp/ttnn/operations/moreh/moreh_fold/device/fold_device_operation.hpp]: Resolved in latest commit

File: ttnn/cpp/ttnn/operations/moreh/moreh_fold/device/fold_program_factory_rm.cpp (hunk 1 above)
Comment (DuongQLee) [reply ttnn/cpp/ttnn/operations/moreh/moreh_fold/device/fold_program_factory_rm.cpp]: Resolved in latest commit

File: ttnn/CMakeLists.txt [hunk 5]
```diff
@@ -505,7 +505,15 @@ set(ALL_TTNN_SRCS
...
     ${CMAKE_CURRENT_SOURCE_DIR}/cpp/ttnn/operations/moreh/moreh_sum/moreh_sum_pybind.cpp
     ${CMAKE_CURRENT_SOURCE_DIR}/cpp/ttnn/operations/moreh/moreh_sum/moreh_sum.cpp
+
+    ${CMAKE_CURRENT_SOURCE_DIR}/cpp/ttnn/operations/moreh/moreh_fold/fold_pybind.cpp
+    ${CMAKE_CURRENT_SOURCE_DIR}/cpp/ttnn/operations/moreh/moreh_fold/fold.cpp
+    ${CMAKE_CURRENT_SOURCE_DIR}/cpp/ttnn/operations/moreh/moreh_fold/device/fold_device_operation.cpp
+    ${CMAKE_CURRENT_SOURCE_DIR}/cpp/ttnn/operations/moreh/moreh_fold/device/fold_program_factory_rm.cpp
+
     ${CMAKE_CURRENT_SOURCE_DIR}/cpp/ttnn/operations/moreh/moreh_helper_functions.cpp
+
```
Comment (hschoi4448) [inline ttnn/CMakeLists.txt:515]: It seems like we need a "format document" feature.

File: ttnn/cpp/ttnn/operations/moreh/moreh_fold/device/kernels/reader_fold_rm.cpp [hunk 6]
```diff
@@ -0,0 +1,109 @@
+// SPDX-FileCopyrightText: © 2023 Tenstorrent Inc.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#include <cstdint>
+
+#include "dataflow_api.h"
+#include "debug/dprint.h"
```
Comment (hschoi4448) [inline ttnn/cpp/ttnn/operations/moreh/moreh_fold/device/kernels/reader_fold_rm.cpp:8]:
```suggestion
```
Please delete it.

File: ttnn/cpp/ttnn/operations/moreh/moreh_fold/device/kernels/reader_fold_rm.cpp [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+        for (uint32_t elem_id = 0; elem_id < W; elem_id++) {
+            uint32_t gid = row_id * W + elem_id;
+            uint32_t nch = gid / W;
+            uint32_t w = gid % W;
+            uint32_t nc = nch / H;
+            uint32_t h = nch % H;
+            uint32_t n = nc / C;
+            uint32_t c = nc % C;
+
+            float sum = 0.0;
```
Comment (hschoi4448) [inline ttnn/cpp/ttnn/operations/moreh/moreh_fold/device/kernels/reader_fold_rm.cpp:53]:
```suggestion
            float sum = 0.0f;
```

File: ttnn/cpp/ttnn/operations/moreh/moreh_fold/device/kernels/reader_fold_rm.cpp [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+                    cb_wait_front(input_cb_id, onetile);
+                    // TODO: Duong double check the DTYPE_BFLOAT16 conversion to float
+                    // This cause precision problems when the input and output is large (a lot of calculations)
+#ifdef DTYPE_BFLOAT16
+                    uint16_t *input_cb_ptr_uint16 = reinterpret_cast<uint16_t *>(get_read_ptr(input_cb_id));
+                    uint16_t bfloat16_value = input_cb_ptr_uint16[lh * LW + lw];
+                    uint32_t float_value_as_int = static_cast<uint32_t>(bfloat16_value) << 16;
+                    auto tmp = reinterpret_cast<float *>(&float_value_as_int);
+                    float value_as_float = *tmp;
+                    sum += value_as_float;
```
Comment (hschoi4448) [inline ttnn/cpp/ttnn/operations/moreh/moreh_fold/device/kernels/reader_fold_rm.cpp:88]: It seems there are two different implementations of `bfloat16_to_float` currently in TT.
Could you check if the results change when using these `bfloat16 to _float` implementations?

1.
```C++
static inline float bfloat16_to_float(uint16_t bfloat_val) {
    uint32_t uint32_data = ((uint32_t)bfloat_val) << 16;
    float f;
    std::memcpy(&f, &uint32_data, sizeof(f));
    return f;
}
```

2.
```C++
float bfloat16_to_float32(uint16_t bfloat16_data) {
    uint32_t bits = static_cast<uint32_t>(bfloat16_data) << 16;

    // Extract the sign bit
    uint32_t sign = bits & 0x80000000;

    // Extract the exponent
    uint32_t exponent = bits & 0x7F800000;

    // Extract the mantissa
    uint32_t mantissa = bits & 0x007FFFFF;

    // Handle special cases
    if (exponent == 0 && mantissa == 0) {
        // Zero
        return sign ? -0.0f : 0.0f;
    } else if (exponent == 0x7F800000) {
        if (mantissa == 0) {
            // Infinity
            return sign ? -__builtin_huge_valf() : __builtin_huge_valf();
        } else {
            // NaN
            return __builtin_nanf("");
        }
    }

    // Assemble the float
    union {
        uint32_t u;
        float f;
    } ieee_float;

    ieee_float.u = sign | exponent | mantissa;
    return ieee_float.f;
}
```

File: ttnn/cpp/ttnn/operations/moreh/moreh_fold/device/kernels/writer_fold_rm.cpp [hunk 9]
```diff
@@ -0,0 +1,30 @@
+// SPDX-FileCopyrightText: © 2023 Tenstorrent Inc.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#include <cstdint>
+
+#include "dataflow_api.h"
+#include "debug/dprint.h"
```
Comment (hschoi4448) [inline ttnn/cpp/ttnn/operations/moreh/moreh_fold/device/kernels/writer_fold_rm.cpp:8]:
```suggestion
```
Please delete it

**Review by thanhnguyen1512 — APPROVED** [review]

File: ttnn/cpp/ttnn/operations/moreh/moreh_fold/device/kernels/reader_fold_rm.cpp (hunk 8 above)
Comment (DuongQLee) [reply ttnn/cpp/ttnn/operations/moreh/moreh_fold/device/kernels/reader_fold_rm.cpp:88]: I use `float_to_bfloat16` to store the sum value to output_cb and I have tested both implementation of bf16 to float you mentioned above, and it still fails big tests. All smaller tests still passed.
```
static inline uint16_t float_to_bfloat16(float val) {
    union {
        float f;
        uint32_t u;
    } ret;
    ret.f = val;
    return uint16_t(ret.u >> 16);
}
```

**Comment by ntarafdar** [comment]
@yan-zaretskiy could you have a look at this since you implemented the OG fold

File: ttnn/cpp/ttnn/operations/moreh/moreh_fold/device/kernels/reader_fold_rm.cpp (hunk 8 above)
Comment (razorback3) [reply ttnn/cpp/ttnn/operations/moreh/moreh_fold/device/kernels/reader_fold_rm.cpp:88]: You mean that unit test of this implementation fails in some cases regardless of which typecasting function you use?

File: tests/tt_eager/ops/test_fold_op.cpp [hunk 10]
```diff
@@ -7,9 +7,9 @@
 #include <random>
```
Comment (mrshaw01) [inline tests/tt_eager/ops/test_fold_op.cpp:7]: Since you didn't make any modifications, please undo the commit for this file.

File: tests/ttnn/unit_tests/operations/test_moreh_fold.py [hunk 11]
```diff
@@ -0,0 +1,108 @@
...
+    [
+        torch.float,
+    ],
+)
+def test_fold_fp32(device, input_shape, output_size, kernel_size, dilation, padding, stride, dtype):
+    torch.manual_seed(2024)
+    run_fold_test(device, input_shape, output_size, kernel_size, dilation, padding, stride, dtype)
+
+
+@skip_for_grayskull("Requires womrhole")
```
Comment (mrshaw01) [inline tests/ttnn/unit_tests/operations/test_moreh_fold.py:56]: Can't this operation pass the test on Grayskull? Why isn't it generalized that way?

File: ttnn/CMakeLists.txt (hunk 5 above)
Comment (mrshaw01) [reply ttnn/CMakeLists.txt:515]: I think you should remove the space lines. @DuongQLee

File: ttnn/cpp/ttnn/operations/moreh/moreh_fold/device/fold_device_operation.cpp [hunk 12]
```diff
@@ -0,0 +1,104 @@
...
+    const operation_attributes_t& operation_attributes, const tensor_args_t& tensor_args) {
+    const auto& output_shape = compute_output_shapes(operation_attributes, tensor_args);
+    auto dtype = tensor_args.input.get_dtype();
+    auto layout = tensor_args.input.get_layout();
+    auto device = tensor_args.input.device();
+    if (tensor_args.output.has_value()) {
+        return tensor_args.output.value();
+    }
+    return create_device_tensor(output_shape, dtype, layout, device, operation_attributes.output_memory_config);
+}
```
Comment (mrshaw01) [inline ttnn/cpp/ttnn/operations/moreh/moreh_fold/device/fold_device_operation.cpp:80]: Optimization suggestion: Move the check for `tensor_args.output.has_value()` to the top to avoid unnecessary computations. 

Also, remove intermediate variable declarations to streamline the code.

File: tests/ttnn/unit_tests/operations/test_moreh_fold.py [hunk 13]
```diff
@@ -0,0 +1,108 @@
+# SPDX-FileCopyrightText: © 2024 Tenstorrent Inc.
```
Comment (mrshaw01) [inline tests/ttnn/unit_tests/operations/test_moreh_fold.py:1]: As you have compute kernel, you should add test for compute kernel config.

File: ttnn/cpp/ttnn/operations/moreh/moreh_fold/device/fold_device_operation.hpp [hunk 14]
```diff
@@ -0,0 +1,76 @@
...
+namespace ttnn::operations::moreh::moreh_fold {
+
+struct MorehFoldOperation {
+    struct operation_attributes_t {
+        const std::vector<uint32_t> output_size;
+        const std::vector<uint32_t> kernel_size;
+        const std::vector<uint32_t> dilation;
+        const std::vector<uint32_t> padding;
+        const std::vector<uint32_t> stride;
+        const MemoryConfig output_memory_config;
```
Comment (mrshaw01) [inline ttnn/cpp/ttnn/operations/moreh/moreh_fold/device/fold_device_operation.hpp:19]: "memory_config" is convention
Should use that name throughout for this operation.

File: ttnn/cpp/ttnn/operations/moreh/moreh_fold/device/fold_program_factory_rm.cpp [hunk 15]
```diff
@@ -0,0 +1,217 @@
...
+    if (fp32_dest_acc_en) {
+        reader_defines["FP32_DEST_ACC_EN"] = 1;
+        writer_defines["FP32_DEST_ACC_EN"] = 1;
+    }
+
+    ////////////////////////////////////////////////////////////////////////////
+    //                      DataMovementKernel SetUp
+    ////////////////////////////////////////////////////////////////////////////
+    bool src_is_dram = input.buffer()->buffer_type() == tt::tt_metal::BufferType::DRAM ? 1 : 0;
+    bool out_is_dram = output.buffer()->buffer_type() == tt::tt_metal::BufferType::DRAM ? 1 : 0;
```
Comment (mrshaw01) [inline ttnn/cpp/ttnn/operations/moreh/moreh_fold/device/fold_program_factory_rm.cpp:126]: 'src' vs 'dst' or 'input' vs 'output' seems more reasonable. In this case, I'd prefer to use 'input' vs 'output.'

File: ttnn/cpp/ttnn/operations/moreh/moreh_fold/device/fold_program_factory_rm.cpp [hunk 16]
```diff
@@ -0,0 +1,217 @@
...
+            runtime_args[0] = input_buffer->address();
+        }
+
+        {
+            auto& runtime_args = GetRuntimeArgs(program, writer_kernel_id, core);
+            runtime_args[0] = output_buffer->address();
+        }
+    }
+}
+}  // namespace ttnn::operations::moreh::moreh_fold
```
Comment (mrshaw01) [inline ttnn/cpp/ttnn/operations/moreh/moreh_fold/device/fold_program_factory_rm.cpp:217]: Should be input_buffer_address = tensor_args.input.buffer()->address()
Then use it in the loop.

File: ttnn/cpp/ttnn/operations/moreh/moreh_fold/device/fold_program_factory_rm.cpp [hunk 17]
```diff
@@ -0,0 +1,217 @@
...
+    }
+
+    return {std::move(program), {reader_kernel_id, writer_kernel_id, cores}};
+}
+
+void MorehFoldOperation::ProgramFactory::override_runtime_arguments(
+    cached_program_t& cached_program,
+    const operation_attributes_t& operation_attributes,
+    const tensor_args_t& tensor_args,
+    tensor_return_value_t& tensor_return_value) {
```
Comment (mrshaw01) [inline ttnn/cpp/ttnn/operations/moreh/moreh_fold/device/fold_program_factory_rm.cpp:196]: Name it "output" instead of tensor_return_value

File: ttnn/cpp/ttnn/operations/data_movement/fold/fold.cpp [hunk 18]
```diff
@@ -4,25 +4,29 @@
```
Comment (mrshaw01) [inline ttnn/cpp/ttnn/operations/data_movement/fold/fold.cpp:4]: Did you change anything in this file? If not, please undo the changes.

**Comment by mrshaw01** [comment]
Many conventions were not applied. You should have done it better.

File: ttnn/cpp/ttnn/operations/data_movement/fold/fold.cpp (hunk 18 above)
Comment (DuongQLee) [reply ttnn/cpp/ttnn/operations/data_movement/fold/fold.cpp:4]: Resolved in last commit.

File: ttnn/cpp/ttnn/operations/moreh/moreh_fold/device/fold_program_factory_rm.cpp (hunk 17 above)
Comment (DuongQLee) [reply ttnn/cpp/ttnn/operations/moreh/moreh_fold/device/fold_program_factory_rm.cpp]: Resolved in last commit.

File: ttnn/cpp/ttnn/operations/moreh/moreh_fold/device/fold_program_factory_rm.cpp (hunk 16 above)
Comment (DuongQLee) [reply ttnn/cpp/ttnn/operations/moreh/moreh_fold/device/fold_program_factory_rm.cpp]: Resolved in last commit.

File: ttnn/cpp/ttnn/operations/moreh/moreh_fold/device/fold_program_factory_rm.cpp (hunk 15 above)
Comment (DuongQLee) [reply ttnn/cpp/ttnn/operations/moreh/moreh_fold/device/fold_program_factory_rm.cpp]: Resolved in last commit.

File: ttnn/cpp/ttnn/operations/moreh/moreh_fold/device/fold_device_operation.cpp [hunk 19]
```diff
@@ -0,0 +1,104 @@
...
+#include "fold_device_operation.hpp"
+
+#include "ttnn/tensor/tensor.hpp"
+#include "ttnn/tensor/types.hpp"
+
+namespace ttnn::operations::moreh::moreh_fold {
+void MorehFoldOperation::validate_inputs(
+    const operation_attributes_t& operation_attributes, const tensor_args_t& tensor_args) {
+    auto& input = tensor_args.input;
+    auto& input_shape = input.get_legacy_shape();
```
[the remark this answers is not in the document]
Comment (DuongQLee) [reply ttnn/cpp/ttnn/operations/moreh/moreh_fold/device/fold_device_operation.cpp]: Resolved in last commit.

File: ttnn/cpp/ttnn/operations/moreh/moreh_fold/device/fold_device_operation.hpp (hunk 14 above)
Comment (DuongQLee) [reply ttnn/cpp/ttnn/operations/moreh/moreh_fold/device/fold_device_operation.hpp]: Resolved in last commit.

File: tests/ttnn/unit_tests/operations/test_moreh_fold.py (hunk 13 above)
Comment (DuongQLee) [reply tests/ttnn/unit_tests/operations/test_moreh_fold.py:1]: Resolved in last commit. I have removed compute kernel config as this op doesn't require compute kernel.

File: ttnn/cpp/ttnn/operations/moreh/moreh_fold/device/fold_device_operation.cpp (hunk 12 above)
Comment (DuongQLee) [reply ttnn/cpp/ttnn/operations/moreh/moreh_fold/device/fold_device_operation.cpp]: Resolved in last commit.

File: tests/ttnn/unit_tests/operations/test_moreh_fold.py (hunk 11 above)
Comment (DuongQLee) [reply tests/ttnn/unit_tests/operations/test_moreh_fold.py]: Resolved in last commit.

File: tests/tt_eager/ops/test_fold_op.cpp (hunk 10 above)
Comment (DuongQLee) [reply tests/tt_eager/ops/test_fold_op.cpp]: Resolved in last commit.

File: ttnn/cpp/ttnn/operations/moreh/moreh_fold/device/kernels/writer_fold_rm.cpp (hunk 9 above)
Comment (DuongQLee) [reply ttnn/cpp/ttnn/operations/moreh/moreh_fold/device/kernels/writer_fold_rm.cpp]: Resolved in last commit.

File: ttnn/cpp/ttnn/operations/moreh/moreh_fold/device/kernels/reader_fold_rm.cpp (hunk 8 above)
Comment (DuongQLee) [reply ttnn/cpp/ttnn/operations/moreh/moreh_fold/device/kernels/reader_fold_rm.cpp:87]: Resolved in the latest commit: I have modified the pytest configuration to ensure that torch.rand generates primarily positive values. This adjustment prevents the cumulative sum of the output from approaching zero, allowing for more flexibility when comparing results using rtol and atol.

File: ttnn/cpp/ttnn/operations/moreh/moreh_fold/device/kernels/reader_fold_rm.cpp (hunk 7 above)
Comment (DuongQLee) [reply ttnn/cpp/ttnn/operations/moreh/moreh_fold/device/kernels/reader_fold_rm.cpp]: Resolved in last commit.

File: ttnn/cpp/ttnn/operations/moreh/moreh_fold/device/kernels/reader_fold_rm.cpp (hunk 6 above)
Comment (DuongQLee) [reply ttnn/cpp/ttnn/operations/moreh/moreh_fold/device/kernels/reader_fold_rm.cpp]: Resolved in last commit.

File: ttnn/CMakeLists.txt (hunk 5 above)
Comment (DuongQLee) [reply ttnn/CMakeLists.txt]: Resolved in last commit.

**Comment by DuongQLee** [comment]
> Many conventions were not applied. You should have done it better.

Resolved in last commit.

File: ttnn/cpp/ttnn/operations/moreh/moreh_fold/device/fold_program_factory_rm.cpp [hunk 20]
```diff
@@ -0,0 +1,203 @@
...
+        {
+            auto& runtime_args = GetRuntimeArgs(program, reader_kernel_id, core);
+            runtime_args[0] = tensor_args.input.buffer()->address();
+        }
+
+        {
+            auto& runtime_args = GetRuntimeArgs(program, writer_kernel_id, core);
+            runtime_args[0] = output.buffer()->address();
+        }
+    }
```
Comment (mrshaw01) [inline ttnn/cpp/ttnn/operations/moreh/moreh_fold/device/fold_program_factory_rm.cpp:201]: Not resolved.

File: tests/ttnn/unit_tests/operations/test_moreh_fold.py [hunk 21]
```diff
@@ -0,0 +1,87 @@
...
+
+# SPDX-License-Identifier: Apache-2.0
+
+import pytest
+import torch
+import ttnn
+
+from loguru import logger
+from models.utility_functions import comp_allclose, comp_allclose_and_pcc, skip_for_grayskull
```
Comment (mrshaw01) [inline tests/ttnn/unit_tests/operations/test_moreh_fold.py:10]: from models.utility_functions import comp_allclose_and_pcc

Comment (DuongQLee) [reply tests/ttnn/unit_tests/operations/test_moreh_fold.py]: Resolved in latest commit.

File: ttnn/cpp/ttnn/operations/moreh/moreh_fold/device/fold_program_factory_rm.cpp (hunk 20 above)
Comment (DuongQLee) [reply ttnn/cpp/ttnn/operations/moreh/moreh_fold/device/fold_program_factory_rm.cpp]: Resolved in latest commit.

**Review by mrshaw01 — APPROVED** [review]

File: tests/ttnn/unit_tests/operations/test_moreh_fold.py [hunk 22]
```diff
@@ -0,0 +1,87 @@
...
+)
+def test_fold(device, input_shape, output_size, kernel_size, dilation, padding, stride, dtype):
+    torch.manual_seed(2024)
+    run_fold_test(device, input_shape, output_size, kernel_size, dilation, padding, stride, dtype)
+
+
+@pytest.mark.parametrize(
+    "input_shape,output_size,kernel_size,dilation,padding,stride",
+    [
+        ((1, 32, 32), (7, 11), (4, 4), (1, 1), (0, 0), (1, 1)),  # input single tile
```
Comment (hschoi4448) [inline tests/ttnn/unit_tests/operations/test_moreh_fold.py:64]:
```suggestion
        [(1, 32, 32), (7, 11), (4, 4), (1, 1), (0, 0), (1, 1)],  # input single tile
```
list is better

File: ttnn/cpp/ttnn/operations/moreh/moreh_fold/device/fold_program_factory_rm.cpp [hunk 23]
```diff
@@ -0,0 +1,205 @@
...
+    uint32_t spatial_dim_size = output_shape.size() - 2;
+    uint32_t kernel_size_product = 1;
+    std::vector<uint32_t> ls;
+    uint32_t L = 1;
+    for (uint32_t i = 0; i < spatial_dim_size; ++i) {
+        uint32_t l =
+            (((operation_attributes.output_size[i] + 2 * operation_attributes.padding[i] -
+               operation_attributes.dilation[i] * (operation_attributes.kernel_size[i] - 1) - 1) /
+              operation_attributes.stride[i]) +
+             1);
```
Comment (hschoi4448) [inline ttnn/cpp/ttnn/operations/moreh/moreh_fold/device/fold_program_factory_rm.cpp:38]:
```suggestion
        uint32_t l =
            (((output_size[i] + 2 * padding[i] -
               dilation[i] * (kernel_size[i] - 1) - 1) /
              stride[i]) +
             1);
```
I think it would be better to modify it this way since the `output_size` variable is created above.

**Comment by hschoi4448** [comment]
It seems that PyTorch's fold supports 2D input, but the tests only show 3D inputs. 
1. Does the current implementation support 2D?
2. is 2D support outside the scope of coverage?

**Comment by DuongQLee** [comment]
> ```python
> import torch
> 
> fold = torch.nn.Fold(output_size=(4, 5), kernel_size=(2, 2))
> input = torch.randn(3 * 2 * 2, 12)
> output = fold(input)
> 
> print(input.shape)
> print(output)
> ```
> 
> It seems that PyTorch's fold supports 2D input, but the tests only show 3D inputs.
> 
> 1. Does the current implementation support 2D?
> 2. is 2D support outside the scope of coverage?

It's my mistake. Currently I have already tested with size (1,32,32), which mean it can support 2D tensor but I only validate 3D tensor. Let me modify the code a bit to fully support both 2D and 3D tensor.

File: ttnn/cpp/ttnn/operations/moreh/moreh_fold/device/fold_program_factory_rm.cpp [hunk 24]
```diff
@@ -0,0 +1,205 @@
...
+    uint32_t dilation_h = dilation[0];
+    uint32_t dilation_w = dilation[1];
+    uint32_t LH = ls[0];
+    uint32_t LW = ls[1];
+
+    ////////////////////////////////////////////////////////////////////////////
+    //                      Device Setup
+    ////////////////////////////////////////////////////////////////////////////
+    Program program{};
+    tt::tt_metal::Device* device = input.device();
```
[the remark this answers is not in the document]
Comment (DuongQLee) [reply ttnn/cpp/ttnn/operations/moreh/moreh_fold/device/fold_program_factory_rm.cpp]: Resolved in lastest commit

File: ttnn/cpp/ttnn/operations/moreh/moreh_fold/device/fold_program_factory_rm.cpp (hunk 23 above)
Comment (DuongQLee) [reply ttnn/cpp/ttnn/operations/moreh/moreh_fold/device/fold_program_factory_rm.cpp]: Resolved in lastest commit

File: tests/ttnn/unit_tests/operations/test_moreh_fold.py (hunk 22 above)
Comment (DuongQLee) [reply tests/ttnn/unit_tests/operations/test_moreh_fold.py:67]: Resolved in lastest commit

File: ttnn/CMakeLists.txt [hunk 25]
```diff
@@ -505,7 +505,13 @@ set(ALL_TTNN_SRCS
     ${CMAKE_CURRENT_SOURCE_DIR}/cpp/ttnn/operations/moreh/moreh_sum/device/moreh_sum_w_program_factory.cpp
     ${CMAKE_CURRENT_SOURCE_DIR}/cpp/ttnn/operations/moreh/moreh_sum/moreh_sum_pybind.cpp
     ${CMAKE_CURRENT_SOURCE_DIR}/cpp/ttnn/operations/moreh/moreh_sum/moreh_sum.cpp
+    ${CMAKE_CURRENT_SOURCE_DIR}/cpp/ttnn/operations/moreh/moreh_fold/fold_pybind.cpp
+    ${CMAKE_CURRENT_SOURCE_DIR}/cpp/ttnn/operations/moreh/moreh_fold/fold.cpp
+    ${CMAKE_CURRENT_SOURCE_DIR}/cpp/ttnn/operations/moreh/moreh_fold/device/fold_device_operation.cpp
+    ${CMAKE_CURRENT_SOURCE_DIR}/cpp/ttnn/operations/moreh/moreh_fold/device/fold_program_factory_rm.cpp
     ${CMAKE_CURRENT_SOURCE_DIR}/cpp/ttnn/operations/moreh/moreh_helper_functions.cpp
+
+
```
[the remark this answers is not in the document]
Comment (DuongQLee) [reply ttnn/CMakeLists.txt]: Resolved in lastest commit

File: ttnn/CMakeLists.txt [hunk 26]
```diff
@@ -505,7 +505,12 @@ set(ALL_TTNN_SRCS
     ${CMAKE_CURRENT_SOURCE_DIR}/cpp/ttnn/operations/moreh/moreh_sum/device/moreh_sum_w_program_factory.cpp
     ${CMAKE_CURRENT_SOURCE_DIR}/cpp/ttnn/operations/moreh/moreh_sum/moreh_sum_pybind.cpp
     ${CMAKE_CURRENT_SOURCE_DIR}/cpp/ttnn/operations/moreh/moreh_sum/moreh_sum.cpp
+    ${CMAKE_CURRENT_SOURCE_DIR}/cpp/ttnn/operations/moreh/moreh_fold/fold_pybind.cpp
+    ${CMAKE_CURRENT_SOURCE_DIR}/cpp/ttnn/operations/moreh/moreh_fold/fold.cpp
+    ${CMAKE_CURRENT_SOURCE_DIR}/cpp/ttnn/operations/moreh/moreh_fold/device/fold_device_operation.cpp
+    ${CMAKE_CURRENT_SOURCE_DIR}/cpp/ttnn/operations/moreh/moreh_fold/device/fold_program_factory_rm.cpp
     ${CMAKE_CURRENT_SOURCE_DIR}/cpp/ttnn/operations/moreh/moreh_helper_functions.cpp
+
```
Comment (hschoi4448) [inline ttnn/CMakeLists.txt:513]:
```suggestion
```
Please remove this.

File: ttnn/cpp/ttnn/operations/moreh/moreh_fold/device/fold_device_operation.cpp [hunk 27]
```diff
@@ -0,0 +1,118 @@
...
+#include "fold_device_operation.hpp"
+
+#include "ttnn/tensor/tensor.hpp"
+#include "ttnn/tensor/types.hpp"
+
+namespace ttnn::operations::moreh::moreh_fold {
+void MorehFoldOperation::validate_inputs(
+    const operation_attributes_t& operation_attributes, const tensor_args_t& tensor_args) {
+    auto& input = tensor_args.input;
+    auto input_shape = input.get_logical_shape().as_vector();
```
Comment (hschoi4448) [inline ttnn/cpp/ttnn/operations/moreh/moreh_fold/device/fold_device_operation.cpp:14]: Wouldn't it be okay to use shape instead of vector?

File: ttnn/cpp/ttnn/operations/moreh/moreh_fold/device/fold_device_operation.cpp [hunk 28]
```diff
@@ -0,0 +1,118 @@
...
+    const std::vector<uint32_t> kernel_size,
+    const std::vector<uint32_t> dilation,
+    const std::vector<uint32_t> padding,
+    const std::vector<uint32_t> stride,
+    const std::optional<MemoryConfig>& memory_config) {
+    return {
+        operation_attributes_t{
+            output_size, kernel_size, dilation, padding, stride, memory_config.value_or(input.memory_config())},
+
+        tensor_args_t{input, output}};
```
Comment (hschoi4448) [inline ttnn/cpp/ttnn/operations/moreh/moreh_fold/device/fold_device_operation.cpp:116]:
```suggestion
        operation_attributes_t{
            output_size, kernel_size, dilation, padding, stride, memory_config.value_or(input.memory_config())},
        tensor_args_t{input, output}};
```
It looks like `clang-format` hasn't been run, given the presence of this extra whitespace.

Comment (DuongQLee) [reply ttnn/cpp/ttnn/operations/moreh/moreh_fold/device/fold_device_operation.cpp:116]: I just ran clang format manually and there is still extra whitespace there. I will remove it then.

File: ttnn/cpp/ttnn/operations/moreh/moreh_fold/device/fold_device_operation.cpp (hunk 27 above)
Comment (DuongQLee) [reply ttnn/cpp/ttnn/operations/moreh/moreh_fold/device/fold_device_operation.cpp]: resolved in latest commit.

File: ttnn/CMakeLists.txt (hunk 26 above)
Comment (DuongQLee) [reply ttnn/CMakeLists.txt]: resolved in latest commit.

File: ttnn/cpp/ttnn/operations/moreh/moreh_fold/device/fold_device_operation.cpp (hunk 28 above)
Comment (DuongQLee) [reply ttnn/cpp/ttnn/operations/moreh/moreh_fold/device/fold_device_operation.cpp]: resolved in latest commit.

**Review by hschoi4448 — APPROVED** [review]

**Review by razorback3 — APPROVED** [review]

**Review by TT-BrianLiu — APPROVED** [review]

**Review by TT-BrianLiu — COMMENTED** [review]
Similar comments to: https://github.com/tenstorrent/tt-metal/pull/13524#discussion_r1811003165

**Comment by DuongQLee** [comment]
> Similar comments to: [#13524 (comment)](https://github.com/tenstorrent/tt-metal/pull/13524#discussion_r1811003165)

I already used grid_to_cores to make the vector cores in program factory.

**Comment by DuongQLee** [comment]
Hello @ayerofieiev-tt @dmakoviichuk-tt @yan-zaretskiy, TT team has finished reviewing my PR. Can you take a look and give me a thumbs up? Thanks

**Review by ayerofieiev-tt — APPROVED** [review]
